optuna/study/_optimize.py

```diff
@@ -189,6 +189,7 @@ def _run_trial(
         optuna.storages.fail_stale_trials(study)
 
     trial = study.ask()
+    study._thread_local.cache_all_trials.activate(lambda: study.get_trials(deepcopy=False))
```

Comment:
This cache will not do anything when optimizing `Study.ask` and `Study.tell` . I think that was one benefit with the current `_CachedStorage`. Is this something that you've thought about?

Reply:
right, this cache will not be available for ask-and-tell users. my (and @c-bata -san)'s plan is to move all cache logic in  `_CachedStorage` to `RDBStorage` because `_CachedStorage` is only used by `RDBStorage`.
Note that this cache is still necessary for users of the other storage such as `JournalRedisStorage`.